File: sdk/python/src/sentinel/webhooks.py
from __future__ import annotations

import hashlib
import hmac
from datetime import UTC, datetime


class WebhookSignatureError(Exception):
    pass


def verify_webhook_signature(
    *,
    header: str,
    secret: str,
    body: bytes,
    now: datetime,
    skew_seconds: int = 300,
) -> None:
    parts = dict(p.split("=", 1) for p in header.split(",") if "=" in p)
    if "t" not in parts or "v1" not in parts:
        raise WebhookSignatureError("malformed signature header")
    try:
        ts = datetime.fromtimestamp(int(parts["t"]), tz=UTC)
    except (ValueError, OverflowError) as exc:
        raise WebhookSignatureError("invalid timestamp") from exc
    if abs((now - ts).total_seconds()) > skew_seconds:
        raise WebhookSignatureError("timestamp outside allowed skew")
    message = f"{int(ts.timestamp())}.".encode() + body
    expected = hmac.new(secret.encode(), message, hashlib.sha256).hexdigest()
    if not hmac.compare_digest(expected, parts["v1"]):
        raise WebhookSignatureError("signature mismatch")

File: sdk/node/src/webhooks.ts
import { createHmac, timingSafeEqual } from "node:crypto";

export class WebhookSignatureError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "WebhookSignatureError";
  }
}

export interface VerifyWebhookSignatureOptions {
  header: string;
  secret: string;
  body: Buffer | Uint8Array | string;
  now: Date;
  skewSeconds?: number;
}

function parseHeader(header: string): { t: number; v1: string } {
  const parts = header.split(",").reduce<Record<string, string>>((acc, item) => {
    const eq = item.indexOf("=");
    if (eq > 0) {
      acc[item.slice(0, eq).trim()] = item.slice(eq + 1).trim();
    }
    return acc;
  }, {});
  const tRaw = parts.t;
  const v1 = parts.v1;
  if (!tRaw || !v1) {
    throw new WebhookSignatureError("malformed signature header");
  }
  const t = Number(tRaw);
  if (!Number.isFinite(t)) {
    throw new WebhookSignatureError("invalid timestamp");
  }
  return { t, v1 };
}

function toBuffer(body: Buffer | Uint8Array | string): Buffer {
  if (Buffer.isBuffer(body)) {
    return body;
  }
  if (typeof body === "string") {
    return Buffer.from(body);
  }
  return Buffer.from(body);
}

export function verifyWebhookSignature(options: VerifyWebhookSignatureOptions): void {
  const skew = options.skewSeconds ?? 300;
  const { t, v1 } = parseHeader(options.header);
  const nowSeconds = Math.floor(options.now.getTime() / 1000);
  if (Math.abs(nowSeconds - t) > skew) {
    throw new WebhookSignatureError("timestamp outside allowed skew");
  }
  const buffer = toBuffer(options.body);
  const message = Buffer.concat([Buffer.from(`${t}.`), buffer]);
  const expected = createHmac("sha256", options.secret).update(message).digest("hex");
  const expectedBuf = Buffer.from(expected, "utf8");
  const providedBuf = Buffer.from(v1, "utf8");
  if (expectedBuf.length !== providedBuf.length || !timingSafeEqual(expectedBuf, providedBuf)) {
    throw new WebhookSignatureError("signature mismatch");
  }
}
